keps/sig-network/2091-cluster-network-policy/README.md

```diff
@@ -0,0 +1,958 @@
+
+#### IPBlock
+
+The `ClusterNetworkPolicyPeer` and `DefaultNetworkPolicyPeer` both allow the
+ability to set an `IPBlock` as a peer. The usage of this field is similar to
+how it is used in the NetworkPolicyPeer. However, we should also explicitly
+note that the IPBlock set in this field could belong to the cluster locally, or
+could be cluster external. For example, the peer could be set with a subnet
+or an IP which maps to a Pod existing in the cluster. This means that a Pod
+in a cluster-scoped NetworkPolicy could be identified by either the labels
+applied on the Pod, or its IP address. In case of multiple conflicting rules
+targeting the same Pod, but identified in different ways, such as via labelSelector
+or via PodIP set in IPBlock, the net effect of the rules will be determined
+by the `action` associated with the rule and/or the resource in which the
+rule is set in, i.e. ClusterNetworkPolicy rule takes precedence over a
+DefaultNetworkPolicy rule.
+
+### Sample Specs for User Stories
+
+![Alt text](user_story_diagram.png?raw=true "User Story Diagram")
+
+#### Story 1: Deny traffic from certain sources
+As a cluster admin, I want to explicitly deny traffic from certain source IPs
+that I know to be bad.
+
+```yaml
+apiVersion: networking.k8s.io/v1alpha1
+kind: ClusterNetworkPolicy
+metadata:
+  name: deny-bad-ip
+spec:
+  appliedTo:
+    # if there's an ingress gateway in the cluster, applying the policy to
+    # gateway namespace will be sufficient
+    namespaceSelector: {}
+  ingress:
+    - action: Deny
+      from:
+      - ipBlock:
+          cidr: 62.210.0.0/16  # blacklisted addresses
+```
+
+#### Story 2: Funnel traffic through ingress/egress gateways
+As a cluster admin, I want to ensure that all traffic coming into (going out of)
+my cluster always goes through my ingress (egress) gateway.
+
+```yaml
+apiVersion: networking.k8s.io/v1alpha1
+kind: ClusterNetworkPolicy
+metadata:
+  name: ingress-egress-gateway
+spec:
+  appliedTo:
+    namespaceSelector:
+      matchLabels:
+        type: tenant  # assuming all tenant namespaces will be created with this label
+  ingress:
+    - action: Deny
+      from:
+      - ipBlock:
+          cidr: 0.0.0.0/0
+      except:
+      - namespaces:
+          self: true
+      - namespaces:
+          selector:
+            matchLabels:
+              kubernetes.io/metadata.name: dmz  # ingress gateway
+  egress:
+    - action: Deny
+      to:
+      - ipBlock:
+          cidr: 0.0.0.0/0
+      except:
+      - namespaces:
+          self: true
+      - namespaces:
+          selector:
+            matchLabels:
+              kubernetes.io/metadata.name: istio-egress  # egress gateway
+```
+
+__Note:__ The above policy is very restrictive, i.e. it rejects ingress/egress
+traffic between tenant Namespaces and `kube-system`. For `coredns` etc. to work,
+`kube-system` Namespace or at least the `coredns` pods needs to be added into
+the Deny `except` list.
```

Comment:
I agree this is indeed a better spec for the usecase. But going off the discussion here https://github.com/kubernetes/enhancements/pull/2522#discussion_r601718058, how would we write such policy if there were no `except` field? Would the cluster admin need to manually put the cluster CIDR in the `ipBlock.except`?

Reply:
Agreed, this use case couldn't easily be described using (the current version of this KEP minus `except`) or even (the current version of this KEP minus `except` plus explicit priority levels).

But there are other ways it might be implemented. eg, if `ipBlock` was understood to only refer to cluster-external IPs, not cluster-internal IPs, then you wouldn't need an `except`. (@tgraf and I were just discussing that...)